File: problems/p445/p445.go
package p445

import "github.com/maomaozgw/leetcode/structures/listnode"

type ListNode = listnode.G[int]

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	l1 = reverse(l1)
	l2 = reverse(l2)

	var head = l1
	var flag = 0
	var lastL1 = head
	for l1 != nil && l2 != nil {
		l1.Val += l2.Val + flag
		flag = l1.Val / 10
		l1.Val %= 10
		lastL1 = l1
		l1 = l1.Next
		l2 = l2.Next
	}
	for l2 != nil {
		lastL1.Next = &ListNode{
			Val: l2.Val + flag,
		}
		l2 = l2.Next
		lastL1 = lastL1.Next
		flag = lastL1.Val / 10
		lastL1.Val %= 10
	}
	for lastL1.Next != nil {
		lastL1.Next.Val += flag
		lastL1 = lastL1.Next
		flag = lastL1.Val / 10
		lastL1.Val %= 10
	}
	if flag > 0 {
		lastL1.Next = &ListNode{
			Val: flag,
		}
	}

	head = reverse(head)

	return head

}

func reverse(head *ListNode) *ListNode {
	var current = head
	var prev *ListNode = nil
	for current != nil {
		var tmp = current.Next
		current.Next = prev
		prev = current
		current = tmp
	}
	return prev
}

File: problems/p445/p445.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        if l1 is None:
            return l2
        elif l2 is None:
            return l1
        l1 = self.reverse(l1)
        l2 = self.reverse(l2)
        flag = 0
        result = ListNode()
        current = result
        while l1 is not None and l2 is not None:
            c = l1.val + l2.val + flag
            if c > 10:
                flag = 1
                c -= 10
            current.next = ListNode(c)
            current = current.next
            l1 = l1.next
            l2 = l2.next
        while l1 is not None:
            c = l1.val = flag
            if c > 10:
                flag = 1
                c -= 10
            current.next = ListNode(c)
            current = current.next
            l1 = l1.next
        while l2 is not None:
            c = l2.val = flag
            if c > 10:
                flag = 1
                c -= 10
            current.next = ListNode(c)
            current = current.next
            l2 = l2.next
        if flag > 0:
            current.next = ListNode(1)
        return self.reverse(result.next)

    def reverse(self, head: ListNode) -> ListNode:
        prev = head
        current = head.next
        prev.next = None
        while current is not None:
            tmp = current.next
            current.next = prev
            prev = current
            current = tmp
        return head
